kythe/cxx/indexer/cxx/clang_range_finder.cc

```diff
@@ -50,35 +44,53 @@ CharSourceRange GetImmediateFileRange(
              : source_manager.getImmediateExpansionRange(loc);
 }
 
-CharSourceRange GetFileRange(const clang::SourceManager& source_manager,
-                             CharSourceRange range) {
+CharSourceRange GetFileRange(CharSourceRange range,
+                             const clang::SourceManager& source_manager,
+                             const clang::LangOptions& lang_options) {
   bool begin_in_macro_body = false;
+  bool begin_at_macro_start = false;
   while (!range.getBegin().isFileID()) {
     begin_in_macro_body = source_manager.isMacroBodyExpansion(range.getBegin());
+    begin_at_macro_start = clang::Lexer::isAtStartOfMacroExpansion(
+        range.getBegin(), source_manager, lang_options);
     range.setBegin(
         GetImmediateFileRange(source_manager, range.getBegin()).getBegin());
   }
   bool end_in_macro_body = false;
+  bool end_at_macro_end = false;
   while (!range.getEnd().isFileID()) {
     end_in_macro_body = source_manager.isMacroBodyExpansion(range.getEnd());
+    end_at_macro_end = clang::Lexer::isAtEndOfMacroExpansion(
+        range.getEnd(), source_manager, lang_options);
     CharSourceRange end = GetImmediateFileRange(source_manager, range.getEnd());
     range.setEnd(end.getEnd());
     range.setTokenRange(end.isTokenRange());
   }
-  if (begin_in_macro_body && end_in_macro_body) {
-    // For macro body ranges we want to return a zero-width range
-    // pointing to the beginning of the macro expansion so that any
-    // edges originating from within the macro body are suppressed.
-    // TODO(shahms): Emit the properly expanded range in all cases once UI
-    // issues are addressed.
+
+  // For macro body ranges we want to return a zero-width range
+  // pointing to the beginning of the macro expansion so that any
+  // edges originating from within the macro body are suppressed.
+  // TODO(shahms): Emit the properly expanded range in all cases once UI
+  // issues are addressed.
+  if (
+      // Only if both the immediately prior range is within a macro body for
+      // both locations.
+      (begin_in_macro_body && end_in_macro_body) &&
+      // And only if both ends would not cover the entire macro expansion.
```

Comment:
Maybe I'm misunderstanding, but wouldn't this rule out, e.g., name decorating macros, like

```c
#define F(X) curse_ ## X

F(you)
```

where the resulting identifier _is_ the whole expansion?

Reply:
I don't know what you mean by "rule out", but this changes the `F(you)` from being undecorated to decorating the entire `F(you)` range.